File: bounded_counter.c
#include <stddef.h>

/*@ 
  requires \valid(counter);
  requires \valid_read(max_value);
  requires *max_value >= 0;
  requires *counter >= 0;
  requires *counter <= *max_value;
  
  assigns *counter;
  
  ensures *counter == (*counter < *max_value) ? *counter + 1 : 0;
  ensures \result == (*counter < *max_value) ? 1 : 0;
*/
int reviewed_bounded_counter(int* counter, const int* max_value) {
    if (counter == NULL || max_value == NULL) {
        return 0;
    }
    
    if (*counter < *max_value) {
        (*counter)++;
        return 1;
    } else {
        *counter = 0;
        return 0;
    }
}

File: BoundedCounter.rs
use prusti_contracts::*;

pub struct BoundedCounter {
    pub value: i32,
}

impl BoundedCounter {
    #[ensures(result.value == 0 && (0 <= result.value) && (result.value <= 5) && (result.value >= 0) && (result.value <= 5))]
    pub fn new() -> Self {
        Self { value: 0 }
    }

    #[pure]
    #[ensures(result == self.value)]
    pub fn get_value(&self) -> i32 {
        self.value
    }

    #[requires((0 <= self.value) && (self.value <= 5) && (self.value >= 0) && (self.value <= 5))]
    #[requires(self.value < 5)]
    #[ensures(self.value == old(self.value) + 1)]
    #[ensures((0 <= self.value) && (self.value <= 5) && (self.value >= 0) && (self.value <= 5))]
    pub fn increment(&mut self) {
        let pre_value = self.value;
        self.value = pre_value + 1;
    }

    #[requires((0 <= self.value) && (self.value <= 5) && (self.value >= 0) && (self.value <= 5))]
    #[requires(self.value > 0)]
    #[ensures(self.value == old(self.value) - 1)]
    #[ensures((0 <= self.value) && (self.value <= 5) && (self.value >= 0) && (self.value <= 5))]
    pub fn decrement(&mut self) {
        let pre_value = self.value;
        self.value = pre_value - 1;
    }
}
